### ISSUE TITLE
amp-analytics Nested Value Template Expansion

### ISSUE THREAD

**Issue by ryanashcraft** [ampproject/amphtml#19548]
## TLDR

I would like to propose a change to the behavior of amp-analytics so that when data is sent via an XHR POST with `useBody` enabled, variable template expansion (aka var substitution) will be performed not just on top-level key-value pairs, but on nested values within the body.

## Higher-Level Goals

I'd like to enable support for vendors like Amplitude to provide an amp-analytics API that:

- Feels ergonomic to the developer instrumenting events with a naturally nested structure, considering the constraints of AMP and of event-based analytics schemas like of Amplitude's and other analytics vendors.
- Feels familiar to the developer if they have instrumented projects with the same vendor with other SDKs and APIs that use a nested structure, so that they can easily map the vendor's general API concepts to the AMP implementation.
- Allows developers to attach useful context available in the various amp variable substitution (e.g. for Amplitude: event properties, user properties, etc.)

## Context and Motivation

See #19160 for some additional context regarding the intent to implement Amplitude as an amp-analytics vendor.

Amplitude's API schema is event-based and supports arbitrary payloads for certain kinds of fields. For example, we accept "event properties", which is a set of key-value pairs that is contextual information for that event. We also accept "user properties" which is a set of key-value pairs contextual for the user. Within each of these, it's up to the user to define their own taxonomy – there's no fixed schema.

Granted there are some constraints, like for example we don't support event/user property values of arbitrary nested depth, we only support array values. The main point is that there's a fixed schema at the top-level.

When starting a prototype for an amp-analytics vendor implementation for Amplitude, I first assessed other vendors in how they designed their API. I found most other vendors either had a more-rigid schema or used certain patterns to workaround the issue. For example, mParticle has separate array fields for all property keys and another for all values. Other vendors flatten the API, but rely on prefixes like "context." or "property." to denote namespaces. My initial prototype followed a similar pattern, where you'd prefix certain fields with namespaces and a colon (e.g. "eventProperties:button id").

Recently, a change was made to amp-analytics to support passing data over an XHR POST via the HTTP request body. This also has enabled analytics vendors to support nested arrays and objects to be passed through the "extraUrlParams" field. Upon coming across this new option, I found like this would help make the Amplitude vendor API feel more ergonomic and familiar for the end-users.

### Ideal API

You can include template variables inside nested object values in extraUrlParams like so:

```
       "vars": {
          "eventType": "click CTA"
       },
       "extraUrlParams": {
          "eventProperties": {
            "page title": "${title}"
          },
          "user_properties": {
            "last page title": "${title}"
          },
          "groups": {
            "company": "Google"
          }
        }
```

## Problems

AMP "extraUrlParams" template variables inside nested arrays and objects are not expanded. Currently template variables only get expanded for top-level string values. So in the example above, "${title}" wouldn't be substituted with the page title.

The existing useBody implementation supports nested objects and arrays, but doesn't do template expansion on any of the nested values. This is one issue of several related to nested objects. For example, image requests don't support nested objects at all – they get stringified to "[object Object]".

## Proposed Fix

Change the logic of extra URL param template expansion to traverse down to expand string values not at the top-level of the extraUrlParams payload.

Note that this does not fix other problems related to nested objects in "extraUrlParams". I propose that if vendors wish to expect nested objects in extraUrlParams, that they disable and provide no support for non-xhrPost+useBody requests.

I propose that we not make any change to the behavior of "extraUrlParamsReplaceMap" or "parentPostMessage".

## Scope

This change should not change the behavior for any other transport option (e.g. for image requests or when useBody is false). This change shouldn't cause any other unintended behavioral change or regressions. It should strictly only change the template string values in nested arrays and objects to be expanded, where today those values don't get expanded.

Any other orthogonal issue related to the recently-exposed "useBody" transport option should be out-of-scope.

**Comment by calebcordry** [comment]
cc/ @jeffjose @zhouyx @lannka @torch2424
